File: routing/astar.c

#include "Python.h"
#include <queue>

static PyObject *AStarError;

/*
Calculates the great circle distance between two points on the surface of the
Earth and returns a result in kilometers.
*/
static double
haversine(double lat1, double lon1, double lat2, double lon2)
{
    double s_dlat, s_dlon, a, c;
    // Convert latitude and longitude to radians before we do anything.
    lat1 = radians(lat1);
    lon1 = radians(lon1);
    lat2 = radians(lat2);
    lon2 = radians(lon2);

    s_dlat = sin((lat2 - lat1) / 2);
    s_dlon = sin((lon2 - lon1) / 2);

    a = s_dlat * s_dlat + s_dlon * s_dlon * cos(lat1) * cos(lat2);
    c = 2 * asin(sqrt(a));
    return c * 6367;
}


/*
def exact_dist(source, dest, graph, coords):
    pred_list = {source : {'dist' : 0}}
    closed_set = set()
    unseen = HeapSet()    # keeps a set and heap structure
    unseen.push(0, source)
    while unseen:
        # vert is a string
        dist, vert = unseen.pop()
        closed_set.add(vert)
        if vert == dest:
            return dist
        for arc, length in graph[vert]:
            if arc in closed_set:
                continue
            new_dist = (pred_list[vert]['dist'] + length)
            if arc not in unseen or new_dist < pred_list[arc]['dist']:
                pred_list[arc] = {'pred' : vert, 'dist' : new_dist}
                if arc not in unseen:
                    est = new_dist + distance(arc, dest, coords)
                    unseen.push(est, arc)
    print "Couldn't find a path between source and dest."
    return None
*/

static PyObject *
astar(PyObject *self, PyObject *args) {
    char *source, *dest;     // the source and destination vertice IDs
    PyObject *graph;         // dict: { vert_id : (arc_id, arc_len) } 
    PyObject *coords;        // dict: { vert_id : (lat, lon) }
    PyObject *pred_list;     // dict: { vert_id : dist_from_origin }
    PyObject *unseen;        // nodes to expand next
    PyObject *unseen_set;    // set: test membership for unseen
    PyObject *closed_set;    // set: nodes already expanded
    if (!PyArg_ParseTuple(args, "ssoo", source, dest, graph, coords)) {
        return NULL;
    }
    return NULL;
}


/* Define Methods */

static PyMethodDef module_methods[] = {
    {"astar", (PyCFunction)astar, METH_VARARGS, " "},
    {NULL}
};

#ifndef PyMODINIT_FUNC  /* declarations for DLL import/export */
#define PyMODINIT_FUNC void
#endif
PyMODINIT_FUNC
initastar(void) 
{
    PyObject* m;
    m = Py_InitModule("astar", module_methods);
    if (m == NULL)
        return;
    
    AStarError = PyErr_NewException("astar.error", NULL, NULL);
    Py_INCREF(AStarError);
    PyModule_AddObject(m, "error", AStarError);
}

File: routing/astar.py


from heapq import heappush, heappop
from animator import GraphSearchAnimator


class AStarAnimator(GraphSearchAnimator):
    """ Draw a graph search. """

    def __init__(self, graph, vertex_coords, landmark_dict):
        GraphSearchAnimator.__init__(self, graph, vertex_coords)
        self.landmark_dict = landmark_dict    # for ALT

    def astar_animation(self, source, dest, heuristic, epsilon=1):
        source, dest = self._find_source_dest(source, dest)
        h_fun = self._heuristic_selector(heuristic)
        h = lambda v: h_fun(v, dest) * epsilon
        return self._animation(source, dest, h)

    def dijkstra_animation(self, source, dest):
        source, dest = self._find_source_dest(source, dest)
        # no heurisitic, so just return zero
        h = lambda v: 0
        return self._animation(source, dest, h)

    def alt_animation(self, source, dest, epsilon=1):
        """ A* Landmark Triangle Inequality: ALT Algorithm """
        source, dest = self._find_source_dest(source, dest)
        h = lambda v: self._alt_heuristic(v, dest) * epsilon
        return self._animation(source, dest, h)

    def _astar(self, source, dest, h):
        if not source or not dest:
            return {}, [] 
        sequence = []
        pred_list = {source : {'dist' : 0, 'pred' : None}}
        closed_set = set()
        unseen = [(0, source)]    # keeps a set and heap structure
        while unseen:
            _, vert = heappop(unseen)
            sequence.append((pred_list[vert]['pred'], [vert]))
            if vert in closed_set:
                continue
            elif vert == dest:
                return sequence[1:], pred_list
            closed_set.add(vert)
            self._relax_vertex(vert, dest, pred_list, unseen, closed_set, h)
        return None    # no valid path found

    def _relax_vertex(self, vert, dest, pred_list, 
                        unseen, closed_set, h):
        for arc, arc_len in self.graph[vert]:
            if arc in closed_set: 
                continue    # disgard nodes that already have optimal paths
            new_dist = pred_list[vert]['dist'] + arc_len
            if arc not in pred_list or new_dist < pred_list[arc]['dist']:
                # the shortest path to the arc changed, record this
                pred_list[arc] = {'pred' : vert, 'dist' : new_dist}
                est = new_dist + h(arc)
                heappush(unseen, (est, arc))
